urlc175f05cb8_mchllbrt_PermCode_tgz-pJ8-permlib_examples_UniversalSeparableJ8/src/permlab/examples/TableauState.java

```diff
@@ -0,0 +1,523 @@
+/*
+ * To change this template, choose Tools | Templates
+ * and open the template in the editor.
+ */
+package permlab.examples;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Random;
+import permlib.PermUtilities;
+
+/**
+ *
+ * @author MichaelAlbert
+ */
+public class TableauState {
+
+    private static int inversions(int[] p) {
+        int result = 0;
+        for (int i = 0; i < p.length; i++) {
+            for (int j = i + 1; j < p.length; j++) {
+                if (p[i] > p[j]) {
+                    result++;
+                }
+            }
+        }
+        return result;
+    }
+
+    private static void checkDeletions(int[] newElements) {
+        int[] deletedCopy = Arrays.copyOf(newElements, newElements.length - 1);
+        boolean hasBadDeletion = false;
+        for (int deletedElement = newElements.length - 1; deletedElement >= 0; deletedElement--) {
+//            System.out.println("Checking deletion " + Arrays.toString(deletedCopy));
+            hasBadDeletion = !isMultiplicityFree(deletedCopy);
+            if (hasBadDeletion || deletedElement == 0) {
+                break;
+            }
+            deletedCopy[deletedElement - 1] = newElements[deletedElement];
+        }
+        if (!hasBadDeletion) {
+            // EUREKA!
+            report(Arrays.toString(newElements));
+        }
+    }
+    private int[] p;
+    private boolean[] pInv;
+    private int[][] tableau;
+    private int nextR;
+    private int nextC;
+    static final int DONE = -1;
+    static final int NO_OP = 0;
+
+    public TableauState(int[] p, int[] tableauColumnLengths) {
+        this.p = Arrays.copyOf(p, p.length);
+        this.tableau = new int[tableauColumnLengths.length][];
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;tableauColumnLengths&#39; (index: 1 | type: int[). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;p&#39;. The same parameter type is validated here for example: urlc175f05cb8_mchllbrt_PermCode_tgz-pJ8-permlib_examples_UniversalSeparableJ8/src/permlab/examples/Av231PlusStuff.java:400. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).